src/io.jl

```diff
@@ -1,39 +1,81 @@
 # TODO: maybe more fancy file format and correctness checking should be done
-
+struct HGF_FORMAT end
+struct JSON_FORMAT end
 
 """
-    hg_save(io::IO, h::Hypergraph)
+    hg_save(io::IO, h::Hypergraph, format::Type{HGF_FORMAT})
 
-Saves a hypergraph `h` to an output stream `io`.
+Saves a hypergraph `h` to an output stream `io` in `hgf` format.
 
 """
-function hg_save(io::IO, h::Hypergraph)
+function hg_save(io::IO, h::Hypergraph, format::Type{HGF_FORMAT})
     println(io, length(h.v2he), " ", length(h.he2v))
     for he in h.he2v
         skeys = sort(collect(keys(he)))
         println(io, join(["$k=$(he[k])" for k in skeys], ' '))
     end
 end
 
+
+"""
+    hg_save(io::IO, h::Hypergraph, format::Type{JSON_FORMAT})
+
+Saves a hypergraph `h` to an output stream `io` in `json` format.
+The `json` in output contains the following information (keys):
+
+* `n` : number of vertices
+* `k` : number of hyperedges
+* `m` : a matrix representation of `h` where rows are vertices and columns are hyperedges
+* `v2he` : mapping vertices to hyperedges
+* `v_meta` : vertices metadata
+* `he_meta` : hyperedges metadata
+
+"""
+function hg_save(io::IO, h::Hypergraph, format::Type{JSON_FORMAT})
+    json_hg = Dict{Symbol, Any}()
+
+    json_hg[:n] = nhv(h)
+    json_hg[:k] = nhe(h)
+
+    json_hg[:m] = JSON3.write(Matrix(h))
+    json_hg[:v2he] = JSON3.write(h. v2he)
+
+    json_hg[:v_meta] = JSON3.write(h.v_meta)
+    json_hg[:he_meta] = JSON3.write(h.he_meta)
+
+    JSON3.write(io, json_hg)
+end
+
+
 """
-    hg_save(fname::AbstractString, h::Hypergraph)
+    hg_save(
+        fname::AbstractString, h::Hypergraph;
+        format::Union{Type{HGF_FORMAT}, Type{JSON_FORMAT_DICT}, Type{JSON_FORMAT_MATRIX}}
+    )
 
-Saves a hypergraph `h` to a file `fname`.
+Saves a hypergraph `h` to a file `fname` in the specified `format`.
 
 """
-hg_save(fname::AbstractString, h::Hypergraph) =
-    open(io -> hg_save(io, h), fname, "w")
+hg_save(
+    fname::AbstractString, h::Hypergraph;
+    format::Union{Type{HGF_FORMAT}, Type{JSON_FORMAT}}=HGF_FORMAT) =
```

Comment:
Provide an `abstract` type to avoid unions like this.
Should be something like:
```
abstract type AbstractHG_Format end
struct HGF_FORMAT <: AbstractHG_Format end
struct JSON_FORMAT <: AbstractHG_Format end
```
and now you can write `format::AbstractHG_Format `
In this way adding new formats has less impact on the code and API

Reply:
this also means that you should dispatch on an instance of a singleton not on type - but again - it is a preferred approach.